### ISSUE TITLE
Different image loaded between test.py and detect.py

### ISSUE THREAD

**Issue by EtienneDSamson** [ultralytics/yolov5#5193]
## ❔Question
Why is the image loaded by the dataloader (create_dataloader() function) with the .yaml file in test.py not the same as the image loaded from LoadImages() in detect.py?

## Additional context
I'm trying to use test.py to detect errors in a custom dataset/bad results from inference, but when trying to compare with results obtained by detect.py, the confidence seemed to always be a bit off (even if every parameter/argument was the same). 

I tried verifying everything, until I completely loaded the image twice from the same path, once with LoadImages() and once through the dataloader. When I compared both images (with torch.allclose() and numpy.subtract()) the images were not the same. Some pixels could have a difference of up to +- 83 (without normalization)! 

Shouldn't loaded images be exactly the same, no matter through which loader / dataloader? Do they not use the same underlying code for resizing and such?

Thanks.

**Comment by glenn-jocher** [comment]
@EtienneDSamson 👋 Hello, thanks for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [val.py](https://github.com/ultralytics/yolov5/blob/master/val.py) in YOLOv5 🚀.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, val.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:


## train.py

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L210-L213 **val_loader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L220-L223
- image size: `640`
- rectangular inference: `False`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `None`

## val.py

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. 
- https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/val.py#L152-L153
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `0.5 * maximum stride`


## detect.py

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/detect.py#L120-L128
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): `0.25` 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): `None`


## YOLOv5 PyTorch Hub Inference

YOLOv5 PyTorch Hub models are`AutoShape()` instances used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)
https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/models/common.py#L276-L282

- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.25 `
- [iou threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L288): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `None`

**Comment by EtienneDSamson** [comment]
Hey @glenn-jocher, thanks for the answer. I understand the differences between each (as well as the need for it, since val.py and train.py need annotations and detect.py doesn't). That being said, shouldn't the same image loaded through either loader be exactly the same if the parameters/arguments given are the same?

I am at the moment getting different confidence values from detect and val, even though all my arguments are exactly the same. Through debugging, I went all the way up to realize the images loaded through the different dataloaders at the very start don't produce the same image, even if given the same path and arguments. 
Using for both: image size: 640, rect:True, conf_thresh:0.25, iou_thresh:0.45, multi-label:False, padding: None.

Is this behaviour normal? Is it because the way they are resized, for example, is not the same?

**Comment by EtienneDSamson** [comment]
Since train.py and detect.py load the image differently, wouldn't this tend to cause disrepancies between the way the model is trained (with data from create_dataloader) and is used to produce inference (with data from LoadImages)?

**Comment by glenn-jocher** [comment]
@EtienneDSamson 👋 hi, thanks for letting us know about this possible problem with YOLOv5 🚀. We've created a few short guidelines below to help users provide what we need in order to get started investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

* ✅ **Minimal** – Use as little code as possible that still produces the same problem
* ✅ **Complete** – Provide **all** parts someone else needs to reproduce your problem in the question itself
* ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

In addition to the above requirements, for [Ultralytics](https://ultralytics.com/) to provide assistance your code should be:

* ✅ **Current** – Verify that your code is up-to-date with current GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been resolved by previous commits.
* ✅ **Unmodified** – Your problem must be reproducible without any modifications to the codebase in this repository. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all of the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) and providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by EtienneDSamson** [comment]
Hey @glenn-jocher, I found the issue with the images. 

It seems `LoadImages` (which is used by detect.py) uses always a `cv2.INTER_LINEAR` as interpolation when resizing in `letterbox()`, while `LoadImagesAndLabels` (which is used by `create_dataloader()` in train and test) has some condition with the value of `r` and `augment` which switchs the interpolation between `cv2.INTER_LINEAR` and `cv2.INTER_AREA` in `load_image()`:

```
LoadImages -> letterbox() (augmentations.py line 118):
img = cv2.resize(img, new_unpad, interpolation=cv2.INTER_AREA)

create_dataloader() -> LoadImagesAndLabels -> load_image() (datasets.py line 673) :
        if r != 1:  # if sizes are not equal
            img = cv2.resize(img, (int(w0 * r), int(h0 * r)),
                             interpolation=cv2.INTER_AREA if r < 1 and not self.augment else cv2.INTER_LINEAR)
```

Is this intended behaviour and I am not understanding well the r and augment parameters, or is this a change that was not reflected between both loaders? It seems counterintuitive to have two loaders load the same image with different interpolations, resulting in different results for inference and such from the getgo.

**Comment by EtienneDSamson** [comment]
I will be glad to create a bug report if in fact it is a problem, I just wanted to make sure before that it wasn't bad parameters causing the problem.

**Comment by EtienneDSamson** [comment]
It seems to me the `load_image()` method should not resize the image (or at least not with a different interpolation), and let the job be done by `letterbox`, which is called right after.

**Comment by glenn-jocher** [comment]
@EtienneDSamson load_image() resizes for caching ops independently of letterbox. But don't worry, if the image is already correctly sized letterbox should not run a second resize:
https://github.com/ultralytics/yolov5/blob/a18b0c36cd4df0d3b9c2623c5dda009c5f281ac9/utils/augmentations.py#L117-L118

cv2.INTER_AREA and cv2.INTER_LINEAR are the same when scaling up, when scaling down AREA performs better but is also slower, which is why we do not use it for training, but do use it for val.py. Whether you want to use it for detect.py is up to your use case and the speed-accuracy compromise that works for you. For best results when scaling down you should apply AREA, for fastest results LINEAR.

**Comment by EtienneDSamson** [comment]
@glenn-jocher Thanks for the info, I understand better why the different interpolations are made. This would explain though the multitude of questions/bug reports of people not understanding why the results for detect and val are not the same. As val is mostly a way to validate how well your network will detect images by comparing detections to actual values, it should be important for detect and val to have the same results when given the same image.
